Objective-C/CBLMutableDictionary.mm

```diff
@@ -166,18 +166,12 @@ - (BOOL) setJSON: (NSString*)json error: (NSError**)error {
         CBLWarnError(Database, @"%@: Failed to convert FLValue to FLDict", self);
         return createError(CBLErrorInvalidJSON, @"Value is not a Dictionary", error);
     }
-    
-    FLDict dict = FLValue_AsDict(result);
-    _dict.clear();
-    
-    FLDictIterator iter;
-    FLDictIterator_Begin(dict, &iter);
-    FLValue value;
-    while (NULL != (value = FLDictIterator_GetValue(&iter))) {
-        id val = FLValue_GetNSObject(value, nil);
-        _dict.set(FLDictIterator_GetKeyString(&iter), [val cbl_toCBLObject]);
-        FLDictIterator_Next(&iter);
+
+    CBL_LOCK(self.sharedLock) {
+        MDict<id> mDict(new cbl::DocContext(), result, true);
+        _dict.initAsCopyOf(mDict, true);
```

Comment:
I haven't been away from this code for long so I'm not sure if this is correct. A few things that need to be checked and confirmed:

1. cbl::parseJSON() returns FLValue which is a pointer to fleece data from parsing the JSON). I have just noticed that the fleece data (FLSliceResult) is not in this method so it could be leaked inside the parseJSON() itself.
2. To replace the Dict inside the _dict, the fleeceData from 1 needs to be retained as a root similarly to what has been done in CBLDocument.
3. One of the purposes of the DocContext is to hold the C4Doc which owns the fleece data of the doc. To hold a Non-Doc fleece data, it's better to create a new type.